fcrepo-kernel-modeshape/src/main/java/org/fcrepo/kernel/modeshape/utils/FedoraTypesUtils.java

```diff
@@ -314,4 +335,79 @@ public static Node getJcrNode(final FedoraResource resource) {
         }
         throw new IllegalArgumentException("FedoraResource is of the wrong type");
     }
+
+    /**
+     * Given a JCR Node, fetch the parent's ldp:insertedContentRelation value, if
+     * one exists.
+     *
+     * @param node the JCR Node
+     * @return the ldp:insertedContentRelation Resource, if one exists.
+     */
+    public static Optional<Resource> ldpInsertedContentProperty(final Node node) {
+        try {
+            if (node.getDepth() > 0) {
+                final Node parent = node.getParent();
+
+                if (parent.hasProperty(LDP_MEMBER_RESOURCE) && parent.isNodeType(LDP_INDIRECT_CONTAINER) &&
+                        parent.hasProperty(LDP_INSERTED_CONTENT_RELATION)) {
+                    return Optional.of(createResource(parent.getProperty(LDP_INSERTED_CONTENT_RELATION).getString()));
+                }
+            }
+        } catch (final RepositoryException ex) {
+            throw new RepositoryRuntimeException(ex);
+        }
+        return empty();
+    }
+
+    /**
+     * Using a JCR session, return a function that maps an RDF Resource to a corresponding property name.
+     *
+     * @param session The JCR session
+     * @return a Function that maps a Resource to an Optional-wrapped String
+     */
+    public static Function<Resource, Optional<String>> resourceToProperty(final Session session) {
+        return resource -> {
+            try {
+                final NamespaceRegistry registry = getNamespaceRegistry(session);
+                return Optional.of(registry.getPrefix(resource.getNameSpace()) + ":" + resource.getLocalName());
+            } catch (final RepositoryException ex) {
+                LOGGER.debug("Could not resolve resource namespace ({}): {}", resource.toString(), ex.getMessage());
+            }
+            return Optional.empty();
+        };
+    }
+
+
+    /**
+     * Update the fedora:lastModified date of the parent's ldp:membershipResource if that node is a direct
+     * or indirect container, provided the LDP constraints are valid.
+     *
+     * @param node The JCR node
+     * @throws RepositoryException if a repository exception occurred
+     */
+    public static void touchLdpMembershipResource(final Node node) throws RepositoryException {
+        if (node.getDepth() > 0) {
+            final Node parent = node.getParent();
```

Comment:
This logic only works as expected if the `parent` is the Direct/Indirect container... which is true if the `node` was created with a `slug`. However, in the common case where no `slug` is used and the UUIDPathMinter creates the hierarchy, the true Direct/Indirect container does not get updated. 
This can be demonstrated in an integration test.